lang/funcs/cidr.go

```diff
@@ -113,6 +113,86 @@ var CidrSubnetFunc = function.New(&function.Spec{
 	},
 })
 
+// CidrSubnetsFunc is similar to CidrSubnetFunc but calculates many consecutive
+// subnet addresses at once, rather than just a single subnet extension.
+var CidrSubnetsFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "prefix",
+			Type: cty.String,
+		},
+	},
+	VarParam: &function.Parameter{
+		Name: "newbits",
+		Type: cty.Number,
+	},
+	Type: function.StaticReturnType(cty.List(cty.String)),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		_, network, err := net.ParseCIDR(args[0].AsString())
+		if err != nil {
+			return cty.UnknownVal(cty.String), function.NewArgErrorf(0, "invalid CIDR expression: %s", err)
+		}
+		startPrefixLen, _ := network.Mask.Size()
+
+		prefixLengthArgs := args[1:]
+		if len(prefixLengthArgs) == 0 {
+			return cty.ListValEmpty(cty.String), nil
+		}
+
+		var firstLength int
+		if err := gocty.FromCtyValue(prefixLengthArgs[0], &firstLength); err != nil {
+			return cty.UnknownVal(cty.String), function.NewArgError(1, err)
+		}
+		firstLength += startPrefixLen
+
+		retVals := make([]cty.Value, len(prefixLengthArgs))
+
+		current, _ := cidr.PreviousSubnet(network, firstLength)
+		for i, lengthArg := range prefixLengthArgs {
+			var length int
+			if err := gocty.FromCtyValue(lengthArg, &length); err != nil {
+				return cty.UnknownVal(cty.String), function.NewArgError(i+1, err)
+			}
+
+			if length < 1 {
+				return cty.UnknownVal(cty.String), function.NewArgErrorf(i+1, "must extend prefix by at least one bit")
+			}
+			// For portability with 32-bit systems where the subnet number
+			// will be a 32-bit int, we only allow extension of 32 bits in
+			// one call even if we're running on a 64-bit machine.
```

Comment:
Why does the `length` value have to be an `int` and not an int64? I'm not sure I understand what output from this needs to be contained within a native integer. There's also the `length += startPrefixLen` immediately after the check which could push the value past `MaxInt32`.

Reply:
I copied this check from the `cidrsubnet` function for consistency, but looking deeper it seems like the main reason it was doing that is that the underlying `cidr` library takes an `int` argument, and so even if we decoded into an `int64` here we'd end up truncating it immediately after anyway.

Addressing that would require changing the underlying `cidr` library to have an additional function that can work with bigger integers, and then we could remove this restriction from both `cidrsubnet` and `cidrsubnets`. But that's a more severe change than I was hoping to make here.

My memory of originally implementing these is hazy, but I think we'd convinced ourselves that this restriction was okay for now because IPv4 addresses can't expand over 32 bits anyway and common IPv6 allocation conventions don't call for prefix growth of over 32 bits in a single step. The typical initial space allocations by regional registries are `/32` at most, and then local registries typically extend those to no more than `/56`. An end-organization with a `/48` prefix (the shortest typical allocation) would then typically create their top-level subnets as `/64` (16 more bits). So a prefix extension of >32 bits would buck IPv6 addressing conventions, even though it would be possible in theory.

Reply:
Oh also, to address the other point: it's not that the `length` value _itself_ would be out of range of the int, but rather the network number calculated for the network internally within the `cidr` package. It chooses a number that fits within the number of _additional_ digits indicated and then masks it into the additional bits created by the prefix extension, and _that_ number is an `int` in the `cidr` package.

So, in principle the `cidr` package could use an `int64` or `big.Int` internally to avoid this, but that isn't universally true today. (There was a recent PR to move it in that direction -- using `big.Int` to allow for extensions of up to 128 bits -- but AFAIK it's not universal across all of the calculation functions and so isn't able to meet this use-case _yet_.)

Reply:
Thanks! All makes sense to me, I didn't do the mental math to think about extensions not typically being more than 32 bits.